Sources/Basic/Lock.swift

```diff
@@ -25,3 +26,65 @@ public struct Lock {
         return try body()
     }
 }
+
+enum ProcessLockError: Swift.Error {
+    case unableToAquireLock(errno: Int32)
+}
+
+/// Provides functionality to aquire a lock on a file via POSIX's flock() method.
+/// It can be used for things like serializing concurrent mutations on a shared resource
+/// by mutiple instances of a process.
+public final class FileLock {
+    /// The name of the lock, used in filename of the lock file.
+    let name: String
+
+    /// The directory where the lock file should be created.
+    let cachePath: AbsolutePath
+
+    /// File descriptor to the lock file.
+    private var fd: Int32 = -1
```

Comment:
It seems that this could be an optional instead of having to use -1 (which is of course common in C because it's all that's available).